### ISSUE TITLE
[member-ordering] support abstract methods

### ISSUE THREAD

**Issue by mgol** [typescript-eslint/typescript-eslint#395]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-instance-method",
          "private-instance-method"
        ]
      }
    ],
  }
}
```

```TS
abstract class TestClass {
  abstract bar(): void;
  foo(): void {}
}
```

**Expected Result**

No error

**Actual Result**

```
  3:3  error  Member foo should be declared before all private instance method definitions  @typescript-eslint/member-ordering
```

**Additional Info**

Abstract methods are not necessarily private methods. There are two issues here:
1. There is no private method in this class so the error message doesn't make sense.
1. Nowhere in the settings is the order between abstract & regular methods defined so it shouldn't be enforced.

This came up when I tried to migrate my TSLint config containing:
```json
    "member-ordering": [
      true,
      "public-before-private",
      "static-before-instance",
      "variables-before-functions"
    ],
```
and I failed due to errors around abstract methods handling.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.5.0` |
| `@typescript-eslint/parser`        | `1.5.0` |
| `TypeScript`                       | `3.3.4000` |
| `ESLint`                           | `5.15.3` |
| `node`                             | `v8.15.1` |
| ~~`npm`~~                              | N/A |
| `yarn`                              | `1.15.2` |

**Comment by bradzacher** [comment]
Abstract methods/props were never implemented in this rule.

https://github.com/typescript-eslint/typescript-eslint/blob/3b28cacd49a1bc61ba0c3f8f76096f6902b92848/packages/eslint-plugin/src/rules/member-ordering.ts#L184-L187

**Comment by mgol** [comment]
@bradzacher 
> Abstract methods/props were never implemented in this rule.

I agree adding support for abstract methods/props would be a new feature but I think treating them as private ones is a bug in the plugin logic. It should ignore method/prop types it doesn't recognize and not misreport them as `private`. What do you think?

**Comment by bradzacher** [comment]
Yeah, it's a bug with the error message.
Though the bug is inherently because it doesn't know specifically that the abstract method is abstract, and instead defaulting to the lowest one in the list it can.

**Comment by David-Else** [comment]
Hi, I am having problems with my abstract classes and weird error messages, is there a workaround to stop the erroneous error messages? For example:
```
21 abstract class GameObject {
22  abstract position: Vector2 = [0, 0];
23  public rotation = 0;
24  public scale: Vector2 = [0, 0];
25  public get x(): number {
26    return this.position[0];
27  }
28  public get y(): number {
29    return this.position[1];
30  }
31 }
```
```
 23:3 warning  Member rotation should be declared before all method definitions  @typescript-eslint/member-ordering
 24:3 warning  Member scale should be declared before all method definitions     @typescript-eslint/member-ordering
```
The only method definitions are the getters, and they are after `rotation` and `scale`, so I am not sure what is happening? My rules are default:
```
   "@typescript-eslint/member-ordering": "warn",
```
If I change the order to:
```
  public rotation = 0;
  public scale: Vector2 = [0, 0];
  abstract position: Vector2 = [0, 0];
```
the error goes away, but `abstract position: Vector2 = [0, 0];` is not a method, could the error message be incorrect?

**Comment by bradzacher** [comment]
as mentioned - the error message is incorrect. There is no handling for abstract members. This causes the error message to be wrong.

**Comment by David-Else** [comment]
This is the first major problem I encountered, hope it can be fixed soon :) 

Sorry I don't have enough skill to help yet, thanks to you and all the community for making typescript-eslint awesome.

**Comment by glen-84** [comment]
Shouldn't this be labelled as a `bug` and not an `enhancement`?

**Comment by bradzacher** [comment]
There is a bug with the error message, sure, but this issue is asking for (intentionally) unimplemented logic to be added (as per https://github.com/typescript-eslint/typescript-eslint/issues/395#issuecomment-480566790), hence enhancement.

**Comment by glen-84** [comment]
But there shouldn't be any error messages just because there happen to be `abstract` properties or methods in the class.

As @mgol mentioned, if you're not ready to add support for abstract members, then this rule should simply ignore them for now.

Setting that aside, couldn't `TSAbstractClassProperty` behave the same as `ClassProperty`, and `TSAbstractMethodDefinition` the same as `MethodDefinition`? That method is only trying to determine `method` vs `constructor` vs `field`, right?

**Comment by bradzacher** [comment]
I mean, it's really just bikeshedding - it doesn't matter if it's classed as a bug or a feature - doesn't change the fact that abstract things were not implemented.

That's a question for people that actually use the rule. I personally use the rule, so I don't have an opinion about how it should sort things.

There are two choices - either `abstract` should be an entirely new set of configs (`'public-abstract-method' | 'private-abstract-method' | ...etc`), or if `abstract` should just be handled as part of the existing configs.

If pushed, I guess I'd lean toward the former, because I assume that the original author's intention was the former (hence they opted to skip implementation in their initial implementation).

Happy to accept a PR with either.

**Comment by glen-84** [comment]
> There are two choices ...

IMO, the first option is the feature request, and the 2nd option is the bug fix.

I agree that option 1 is the better option, as it would allow developers to position abstract members differently to non-abstract members, which could result in more consistent ordering.

Looking at the code, could `abstract` be considered another `scope`, or would the naming need to be changed?

**Comment by bradzacher** [comment]
I believe it could be treated as a new scope, i.e same as instance/static.

**Comment by hsmitty93** [comment]
I created a PR to add support of abstract members to the member-ordering rule, would love a review please.  #1004

**Comment by carcade** [comment]
```ts
abstract class Foo {
    public abstract name: string;

    constructor() {}

    public bar(): void {};
}
```

This code with default member-ordering rule config (`"@typescript-eslint/member-ordering": ["error"]`) gives me the following error: 

```
ESLint: Member constructor should be declared before all method definitions. (@typescript-eslint/member-ordering)
```

But if I remove abstract property `name`, there is no such error.

Is it a bug?
